### PR TITLE
First version of ApodiniAudit

### PR DESCRIPTION
**Pull request by simonbohnen**
# First version of ApodiniAudit

## :gear: Release Notes 
This is the first version of ApodiniAudit, which enables developers to automatically test their APIs against common best practices for REST and HTTP. It introduces the `RESTAuditor` and `HTTPAuditor` Configurations which register the `audit` CLI command. This command performs the audit, currently checking for simple best practices such as adequate length of URL path segments.

### Changes outside ApodiniAudit
The concept of `DependentStaticConfigurations` was generalized. The `HTTPExporter` now accepts dependent configurations too, and the `REST` and `OpenAPI` configurations had to be adapted slightly to fit the new model.

### Testing
I am going to implement testing for all the best practices at a later stage.

### Reviewer Nudging
`AuditStaticConfiguration` is a good entry point. It registers the `AuditInterfaceExporter`, which manages the auditing process, and is available as a static configuration for the `HTTP` and `REST` exporters.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/Apodini/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/Apodini/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/Apodini/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/Apodini/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift
Comment (PSchmiedmayer) [inline Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift:28]: I think using variadic parameters and in generally `AnyDependentStaticConfiguration` works in a way that one can pass `ConfigurationWithDependents` to the builder but not only restricting them to a specific configuration type by constraining it to a specific `InternalConfiguration`.
I think that `ParentConfiguration` does not need to conform to `ConfigurationWithDependents` and that the constraint should then be used with concrete `ConfigurationWithDependents` in the build functions using the Swift `where` clause. Otherwise one can pass any `ConfigurationWithDependents` to the builder from the outside and we have the same level of type safety as we have internally using the type erased version.

```swift
@resultBuilder
public struct DependentStaticConfigurationBuilder<Configuration> {
    public static func buildIf<T: ConfigurationWithDependents>(_ component: T?) -> T? where T.Configuration == Configuration {
        return component
    }
    
    public static func buildBlock<T: ConfigurationWithDependents>(_ arg0: T) -> T where T.Configuration == Configuration {
        return arg0
    }

    public static func buildEither<T: ConfigurationWithDependents>(first: T) -> T where T.Configuration == Configuration {
        return first
    }
    
    public static func buildEither<T: ConfigurationWithDependents>(second: T) -> T where T.Configuration == Configuration {
        return second
    }

    public static func buildIf<T: ConfigurationWithDependents>(first: T) -> T where T.Configuration == Configuration {
        return first
    }

    public static func buildBlock<T0: ConfigurationWithDependents, T1: ConfigurationWithDependents >(_ arg0: T0, _ arg1: T1) -> [AnyDependentStaticConfiguration] where T0.Configuration == Configuration, T1.Configuration == Configuration {
        return [arg0, arg1]
    }
}
```
We will need more buildBlock functions with more types, as [Swift sadly does not yet support variadic generics (but this is on the way)](https://forums.swift.org/t/variadic-generics/54511)

File: Sources/ApodiniAudit/BestPractices/BestPractice.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniAudit/BestPractices/BestPractice.swift]: Probably an [OptionSet](https://developer.apple.com/documentation/swift/optionset) might make sense for the category and especially for the scope. Some best practices might lie in multiple categories and the scope might be across multiple exporter types.

File: Sources/ApodiniHTTP/InterfaceExporter.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniHTTP/InterfaceExporter.swift]: As noted above, as far as I can see, the constraint currently doesn't have any effect here.

File: Sources/ApodiniHTTP/InterfaceExporter.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniHTTP/InterfaceExporter.swift:22]: Do the static configurations have to be public? What would be the reason for making them available to the outside? Same applies to the REST exporter.

File: Sources/ApodiniREST/RESTInterfaceExporter.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniREST/RESTInterfaceExporter.swift]: Please remove comments before merging the PR

File: TestWebService/Sources/TestWebService/TestWebService.swift
Comment (PSchmiedmayer) [inline TestWebService/Sources/TestWebService/TestWebService.swift]: I would suggest to name the auditors the same for all exporters. As noted on chat.tum.de: I think it makes sense to even use the same `InternalConfiguration` for the REST and HTTP Exporters.

File: Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift
Comment (simonbohnen) [reply Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift:28]: resultBuilders allow for the conversion to an intermediate type using the `buildExpression` method, as explained [here](https://docs.swift.org/swift-book/ReferenceManual/Attributes.html#ID634). All subsequent methods can take `[AnyDependentStaticConfiguration]`s, as the correct type was already ensured by `buildExpression`. This can be verified by trying to pass an `OpenAPI` Configuration into a `HTTP` Configuration, which will not compile. Swift is actually being quite handy with types here for once!

File: Sources/ApodiniHTTP/InterfaceExporter.swift
Comment (simonbohnen) [reply Sources/ApodiniHTTP/InterfaceExporter.swift:22]: As `ConfigurationWithDependents` is defined in `Apodini`, it has to be public to be accessible from `ApodiniHTTP` and `ApodiniREST`. Swift seems to force me to make the property public too unfortunately. Removing the property from the protocol doesn't seem like a good option, because it is being used in the protocol's extension (see [here](https://github.com/Apodini/Apodini/pull/433/files#diff-5cb1ee72a2e05fe5677b3af34c6d22f81b964f8a72a53c562f2a656719336129)).

File: TestWebService/Sources/TestWebService/TestWebService.swift
Comment (simonbohnen) [reply TestWebService/Sources/TestWebService/TestWebService.swift]: I've merged the two `InternalConfigurations`, which were already the same anyways.

Currently, the type parameter to a `DependentStaticConfiguration` is the outward facing `REST` or `HTTP` configuration, not the internal configuration. Thus, merging the `HTTPAuditor` and `RESTAuditor` is currently not possible, as this would require conforming to the same protocol twice.

If we were to take the internal configuration as a type parameter to `DependentStaticConfiguration` (which was your suggestion anyways @PSchmiedmayer 😅), then that would force us to make the `OpenAPI` configuration available as a dependent of the `HTTP` as well as the `REST` configuration. That's fine by me, and @PSchmiedmayer suggested it too.

Do we need to make any modifications to the OpenAPI exporter for it to work as a dependent of the HTTP exporter? Who would I best talk to for this @PSchmiedmayer?

File: Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift
Comment (PSchmiedmayer) [reply Sources/Apodini/Configurations/DependentStaticConfigurationBuilder.swift:28]: Nice, learned something new, very cool! Thanks for sharing the link and the insight 🚀

File: TestWebService/Sources/TestWebService/TestWebService.swift
Comment (PSchmiedmayer) [reply TestWebService/Sources/TestWebService/TestWebService.swift]: Thanks for the response. I took a look at the diff and the current state and think that making the `OpenAPI` configuration available as a dependent configuration of the `HTTP` and the `REST` Exporter is definitely a good idea and would not require too many changes to the Exporter as far as I can see. 

I see two main differences that would need to be addressed:
1. The response type of an endpoint: The REST Exporter wraps a response in the `ResponseContainer` to add the HATEOAS information to the responses, which the HTTP Exporter does not do. Therefore it would make sense to add the way a response is packaged at the end to the `HTTPExporterConfiguration`. The REST Exporter would share with the OpenAPI Exporter that it uses the `ResponseContainer`, and the HTTP Exporter would share that it doesn't use a wrapping. Moving the `ResponseContainer` type to the ApodiniHTTPProtocol target would then probably be a reasonable idea.
2. The HTTP Exporter should Exporter all request-response Handlers the same way as the REST Exporter does. The exporters currently use different mechanisms to register the endpoints but should result in the same final product. That this is the same could only be guaranteed if they would share the code for registering the routes, e.g., the ApodiniHTTPProtocol target. The only area where this registration differs is in the streaming cases for Handlers (e.g., client-side stream). I am not sure if and how OpenAPI could reflect HTTP/2 streaming behavior, but this might then require some additional code in the OpenAPI Exporter. But this feature is currently still on the roadmap and not implemented, so it would not require immediate changes in the Exporter at the moment, and when you tackle this, we can see how and if this can be incorporated into the OpenAPI Exporter 👍

Comment (simonbohnen) [reply TestWebService/Sources/TestWebService/TestWebService.swift]: That sounds good! Point 1 makes a lot of sense.

Would you include moving the route registration to `ApodiniHTTPProtocol` in this pull request? It seems like that is a bigger task which might deserve its own PR to me.

Comment (simonbohnen) [reply TestWebService/Sources/TestWebService/TestWebService.swift]: I implemented the correct schema exporting for HTTP with OpenAPI! It looked to me like the modifications I made were the only ones necessary, right?

File: Sources/ApodiniAudit/InterfaceExporter/AuditStaticConfiguration.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniAudit/InterfaceExporter/AuditStaticConfiguration.swift]: I think it would be needed to make the mode dependent on the `HTTPExporterConfiguration`.

File: Sources/ApodiniAudit/InterfaceExporter/AuditInterfaceExporter.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniAudit/InterfaceExporter/AuditInterfaceExporter.swift]: Would it be possible to replace this with the `HTTPExporterConfiguration` instead of a separate AuditMode enum as both could express the same functionality and potentially more?

File: Sources/ApodiniAudit/InterfaceExporter/AuditStaticConfiguration.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniAudit/InterfaceExporter/AuditStaticConfiguration.swift]: I would suggest to make the types and the functions static functions of the `APIAuditorConfiguration` to give them sone sort of scope and not add too many global functions and properties.

File: Sources/ApodiniOpenAPI/Document/OpenAPIComponentsObjectBuilder.swift
Comment (PSchmiedmayer) [inline Sources/ApodiniOpenAPI/Document/OpenAPIComponentsObjectBuilder.swift]: Nice, great to see that the changes to make the OpenAPI exporter support the HTTP export were quite small and easy to do 🚀

**Review by PSchmiedmayer — APPROVED** [review]
Thanks for the first Apodini Auditor PR! The changes look good, I only had a few small comments. This PR can be merged once these are addressed, the test code coverage is increased, and the SwiftLint errors are resolved.
Thanks for the work that went into the PR 🚀

**Pull request merged**
